spec.bs

```diff
@@ -491,12 +494,243 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
+  1. Run [=get the next code point=] given |tokenizer|.
+</div>
+
+<div algorithm>
+  To <dfn>add a token</dfn> for a given [=tokenizer=] |tokenizer|, [=token/type=] |type|, number |next position|, number |value position|, and number |value length|:
+
+  1. Let |token| be a new [=token=].
+  1. Set |token|'s [=token/type=] to |type|.
+  1. Set |token|'s [=token/index=] to |tokenizer|'s [=tokenizer/index=].
+  1. Set |token|'s [=token/value=] to the substring within |tokenizer|'s [=tokenizer/input=] starting at |value position| with |value length|.
+  1. Set |tokenizer|'s [=token/index=] to |next position|.
+</div>
+
+<div algorthm>
+  To <dfn>add a token with default length</dfn> for a given [=tokenizer=] |tokenizer|, [=token/type=] |type|, number |next position|, and number |value position|.
+
+  1. Let |computed length| be |next position| &minus; |value position|.
+  1. Run [=add a token=] given |tokenizer|, |type|, |next position|, |value position|, and |computed length|.
+</div>
+
+<div algorithm>
+  To <dfn>add a token with default position and length</dfn> for a given [=tokenizer=] |tokenizer| and [=token/type=] |type|:
+
+  1. Run [=add a token with default length=] given |tokenizer|, |type|, |tokenizer|'s [=tokenizer/next index=], and |tokenizer|'s [=tokenizer/index=].
+</div>
+
+<div algorithm>
+  To <dfn>process a tokenizing error</dfn> for a given [=tokenizer=] |tokenizer|, a number |next position|, and a number |value position|:
+
+  1. If |tokenizer|'s [=tokenizer/policy=] is "[=tokenize policy/strict=]", then throw a {{TypeError}}.
+  1. [=Assert=]: |tokenizer|'s [=tokenizer/policy=] is "[=tokenize policy/lenient=]".
+  1. Run [=add a token with default length=] given |tokenizer|, "[=token/type/invalid-char=]", |next position|, and |value position|.
+</div>
+
+<div algorithm>
+  To <dfn>process a tokenizing error with default position</dfn> for a given [=tokenizer=] |tokenizer|:
+
+  1. Run [=process a tokenizing error=] given |tokenizer|, |tokenizer|'s [=tokenizer/next index=], and |tokenizer|'s [=tokenizer/index=].
+</div>
+
+<div algorithm>
+  To perform <dfn>is a valid name code point</dfn> given a unicode |code point| and a boolean |first|:
+
+  1. If |first| is true return the result of checking if |code point| is contained in the [=IdentifierStart=] set of code points.
+  1. Else return the result of checking if |code point| is contained in the [=IdentifierPart=] set of code points.
+</div>
+
+<div algorithm>
+  To determine if a unicode |code point| <dfn>is ASCII</dfn>:
+
+  1. If |code point| is between U+0000 and U+007F inclusive, then return true.
+  1. Otherwise return false.
+</div>
+
+<div algorithm>
+  To <dfn export>tokenize</dfn> a given [=/pattern string=] |input| and [=tokenize policy=] |policy|:
+
+  1. Let |tokenizer| be a new [=tokenizer=].
+  1. Set |tokenizer|'s [=tokenizer/input=] to |input|.
+  1. Set |tokenizer|'s [=tokenizer/policy=] to |policy|.
+  1. Let |input length| be [$Get$](|input|, "`length`").
```

Comment:
You are not working with JS objects here so you should just say `|input|'s [=string/length=]`.